### DISCUSSION TITLE
Control P-I-D library for esp32

### DISCUSSION THREAD

**Discussion by 2dof** [micropython/micropython#10975] [Libraries & Drivers]
I published small P-I-D control library in micropython for esp32 ( for other platforms should work too):
[ https://github.com/2dof/esp_control](https://github.com/2dof/esp_control)

It is based typical on functional code (as base to OOP) and provide PID algorithms and signal processing based on industrial standard implementation ( but without alarms). 
Also I added a curve generator for Setpoint generation in thermal proces control.

from my test and simulations library should be used for proces samplins Ts > 0.1 sec. 
In the furture I will add other algorithms (auto tuning mechanism, pole placement etc) and  'user-end-implementations' examples. 

I hope description is clear and someone will find it usefull.

**Comment by peterhinch** [comment]
Impressive. And properly documented :+1:

**Comment by 2dof** [comment]
I have added  small lib for Thermocouple type K  functions ( based on Polynomials and lookup tables ) for non-linearity correction:
[thermocouple type K ](https://github.com/2dof/esp_control/tree/main/src/thermocouples).

It is good base for:
- AD8495 / AD8497 linearity correction using NIST Thermoelectric Voltage Lookup Tables. 
- [AD7793+ADT7320 Measurement Solution] (https://www.analog.com/en/analog-dialogue/articles/measuring-temp-using-thermocouples.html)  

Other types  will be added later.

**Reply by ctimmer** [reply]
Impressive indeed!
I'm having trouble setting up a minimal implementation to control an electric oven.

Are there any simple examples using mostly default settings that I could use to get started and then build on as my PID skills improve? It looks like the information is there, I'm looking for something basic that brings it all together.

Please keep up the good work.

Thanks,
Curt

**Reply by 2dof** [reply]
I You are New in Control theory I recommend for You to get acquainted with:
 - > [ProportionalIntegralContro](https://apmonitor.com/pdc/index.php/Main/ProportionalIntegralControl)  -> it has good introduction for   beginners ( + examples [ArduinoTemperatureControl](http://apmonitor.com/pdc/index.php/Main/ArduinoTemperatureControl) , at leat You can learn how to design/simulate without heavy theory. 
- > [Making my own PID library](https://github.com/orgs/micropython/discussions/10180)  user ajevticc  implemented temp. control for his own micropython project. - it good start to "how to" and solving problem.
 
about My lib, ( if You will be using my implementation) I recommend to use  implementation from section [2.2 PID with anti-windup](https://github.com/2dof/esp_control#22-pid-with-anti-windup)  - they a simpler ( than PID-ISA)  and on the end of week w will push 
OOP example (thermal process) as basic template for practical implementation.  

You can follow this discussion because these PID libs  will be base for practical full implementation of controller (at least for "computer in the loop) with auto-tunnig (connect and forget)  simulation example based on ceramic kiln Temp. control model. 

For now I'm not focusing on any hardware API ( LCD, keyboard, menu system helpful to build your own hardware [like  this](https://industrial.omron.co.za/en/products/e5cb), because i want create API for android BT/web server app first , but the are a lot of users on forum that publish their own solutions/drivers  and I will be reference to them.

I You going to build oven from scratch then I recommend use solutions of  actuator ( specialized relays, triacs ) from commercial products
and focus on your code/application.  You can create your own topic and forum users will help You to solve problems.

**Reply by ctimmer** [reply]
Thanks for the feedback, very helpful. I will be watching for your examples.

Curt

**Reply by 2dof** [reply]
@ctimmer   ..I forgot update file , now:
file in [pid_aw] https://github.com/2dof/esp_control/tree/main/src

basic usage: 
1. copy file to your device
2. copy below code to  to  your device or locally to your IDE (Thonny etc.) and run,

Step responce of PID:
```python 

from pid_aw import *

# pid structure creation
pid_buf=bytearray(101)   # size of PID_REGS is 101 bytes, 
PID1 = uctypes.struct(uctypes.addressof(pid_buf), PID_REGS, uctypes.LITTLE_ENDIAN)
pid_init0(PID1)

# lets change some params 
PID1.Kp = 2
PID1.Tt = PID1.Ts            # 0.1
pid_tune(PID1)            

# set P-I D:
PID1.CFG.Psel=  True ; PID1.CFG.Isel = True ;  PID1.CFG.Dsel = True   
# or pid.CFG_REG=0x03


    
    # and test settings
PID1.CFG.Modesel = False # True
     
# get step responce  of PID e =sp-pv = 10 
sp =10.
pv=0.
for i in range(0,25):
    #u = pid_aw_updateControl(PID1,sp,pv,ubias = 0.)
    u = pid_awm_updateControl(PID1,sp,pv,ubias=0.,mv =0.)
    
    print("u:",u,"uk:",PID1.uk)    # PID1.uk is non limited cv (before limitation)
```

In this example  a  pid_init0(pid) function is used to load basic setting.  For control thermal process like oven there time constants are at leastcouple of seconds the sampling time (Ts)  -> PID1.Ts can be larger than now (0.1) , por example 0.5 or 1 sec. ( at least 5 to 10 times less than
time constant of oven).

the basic settings to start up play with oven:

1. set pid control to a P-I controller : 'PID1.CFG.Psel.True =  True ; PID1.CFG.Isel = True ;  PID1.CFG.Dsel = False 
2. then using pid_aw algorithms, remember adjust Tm (= Ts) if neded. 
3. set your own parameters P-I-(D) parameters  KP, Ti (Td, Tm) , in practice You need to know about your thermal process: 
    Gain, and Time constant -> You can perform a "step response of process" and note temp value every couple of seconds, after that you can estimate parameters calculate The PID parameters from Ziegler-Nichols Rules as your first "shot".  
note: most ( for baking) , of the ovens uses "on-off" control, i.e  full power to the heater other (more fancy/specialized, use continuos ( Proportional Control: power from 0 to 100%)  based on triacs  etc. 
depending on Your solution You need remember what a "step"  value will be.

**Reply by ctimmer** [reply]
@2dof 
Thanks so much, this is just what I was looking for. My project is a modified electric smoker with an SSR for proportional control power.

Curt

**Reply by 2dof** [reply]
"SSR for proportional control power" so it is standard ON/OFF control mod (like thermostat)  amd if it is not other "fancy" on-off ( cycle mode or phase mode) You do not even need a PID controller, just simple  relay (on-off) for Your SSR.

**Reply by ctimmer** [reply]
Using an [SSR (see Operation section)](https://en.wikipedia.org/wiki/Solid-state_relay) works much better than a mechanical relay for switching AC power. A PID controller is still needed to set the duty cycle. I have my own algorithm for what is sometimes called slow PWM which works well with AC.

**Reply by 2dof** [reply]
@ctimmer    [Example 2: Class controller example - implementing OOP tutorial](https://github.com/2dof/esp_control/tree/main/Examples) 
maybe will be helpful  for You.

**Comment by 2dof** [comment]
added:

1.  [On-Off controller](https://github.com/2dof/esp_control#23-on-off-controller)   - simple on-off control based on relay 
2.) [ type J thermocouples] (https://github.com/2dof/esp_control/tree/main/src/thermocouples#2-thermocouple-type-j)  library based on its-90, but  lockup tables calculate only in range 0 - 760 Celsjus ( from practical reason). 
3)  Added [Example 2: Class controller example -  implementing OOP tutorial](https://github.com/2dof/esp_control/tree/main/Examples) , as example  using  pid functions in OOP.  Part 1:  class creating and pure soft simulation (Part 2 (in preparation) , using interrupts for practical use / template for example Hardware solution)

+ reloaded pid_aw.py  -> follow the updates

**Reply by 2dof** [reply]
Some update about On-Off control: I have added in [Examples](https://github.com/2dof/esp_control/tree/main/Examples):
[Example_python_FOPDT_on_off.py](https://github.com/2dof/esp_control/blob/main/Examples/Example_python_FOPDT_on_off.py) - simulation script in python  (same code as micropython version), also moved
[on_off_control_example.py](https://github.com/2dof/esp_control/blob/main/Examples/on_off_control_example.py) - simulation in micropython, o examples.
Also, I have added some comments to example about using on-off controller in follow-up control  (and some comment about using it in electric ceramic kiln (simple) temperature control).
